File: packages/sdkwork-utils-csharp/CurrencyUtils.cs
namespace Sdkwork.Utils;

public static class CurrencyUtils
{
    private sealed record CurrencyMeta(int Exponent, string Symbol);

    private static readonly Dictionary<string, CurrencyMeta> Known = new(StringComparer.OrdinalIgnoreCase)
    {
        ["USD"] = new(2, "$"),
        ["EUR"] = new(2, "€"),
        ["GBP"] = new(2, "£"),
        ["CNY"] = new(2, "¥"),
        ["JPY"] = new(0, "¥"),
        ["KRW"] = new(0, "₩"),
        ["HKD"] = new(2, "HK$"),
        ["TWD"] = new(2, "NT$"),
        ["CHF"] = new(2, "CHF"),
        ["CAD"] = new(2, "CA$"),
        ["AUD"] = new(2, "A$"),
        ["INR"] = new(2, "₹"),
        ["BHD"] = new(3, "BHD"),
        ["KWD"] = new(3, "KWD"),
    };

    private static CurrencyMeta? Lookup(string code)
    {
        var normalized = code.Trim();
        if (normalized.Length != 3 || normalized != normalized.ToUpperInvariant()
            || !normalized.All(char.IsLetter))
        {
            return null;
        }
        return Known.TryGetValue(normalized, out var meta) ? meta : null;
    }

    public static bool IsCurrencyCode(string value) => Lookup(value) != null;

    public static int? MinorUnitExponent(string code) => Lookup(code)?.Exponent;

    public static long? ToMinorUnits(double amount, string code)
    {
        var meta = Lookup(code);
        if (meta == null)
        {
            return null;
        }
        var factor = Math.Pow(10, meta.Exponent);
        return (long)NumberUtils.Round(amount * factor, 0);
    }

    public static double? FromMinorUnits(long minor, string code)
    {
        var meta = Lookup(code);
        if (meta == null)
        {
            return null;
        }
        var factor = Math.Pow(10, meta.Exponent);
        return minor / factor;
    }

    private static bool SuffixLocale(string locale)
    {
        var normalized = locale.ToLowerInvariant();
        return normalized.StartsWith("de") || normalized.StartsWith("fr") ||
               normalized.StartsWith("it") || normalized.StartsWith("es");
    }

    public static string? FormatCurrency(double amount, string code, string locale)
    {
        var meta = Lookup(code);
        if (meta == null)
        {
            return null;
        }
        var formatted = I18nUtils.FormatNumberLocale(amount, locale, meta.Exponent);
        return SuffixLocale(locale) ? $"{formatted} {meta.Symbol}" : $"{meta.Symbol}{formatted}";
    }
}

File: packages/sdkwork-utils-kotlin/src/main/kotlin/com/sdkwork/utils/CurrencyUtils.kt
package com.sdkwork.utils

object CurrencyUtils {
    private data class CurrencyMeta(val exponent: Int, val symbol: String)

    private val known = mapOf(
        "USD" to CurrencyMeta(2, "$"),
        "EUR" to CurrencyMeta(2, "€"),
        "GBP" to CurrencyMeta(2, "£"),
        "CNY" to CurrencyMeta(2, "¥"),
        "JPY" to CurrencyMeta(0, "¥"),
        "KRW" to CurrencyMeta(0, "₩"),
        "HKD" to CurrencyMeta(2, "HK$"),
        "TWD" to CurrencyMeta(2, "NT$"),
        "CHF" to CurrencyMeta(2, "CHF"),
        "CAD" to CurrencyMeta(2, "CA$"),
        "AUD" to CurrencyMeta(2, "A$"),
        "INR" to CurrencyMeta(2, "₹"),
        "BHD" to CurrencyMeta(3, "BHD"),
        "KWD" to CurrencyMeta(3, "KWD"),
    )

    private fun lookup(code: String): CurrencyMeta? {
        val normalized = code.trim()
        if (normalized.length != 3 || normalized != normalized.uppercase() || !normalized.all { it.isLetter() }) {
            return null
        }
        return known[normalized]
    }

    fun isCurrencyCode(value: String): Boolean = lookup(value) != null

    fun minorUnitExponent(code: String): Int? = lookup(code)?.exponent

    fun toMinorUnits(amount: Double, code: String): Long? {
        val meta = lookup(code) ?: return null
        val factor = Math.pow(10.0, meta.exponent.toDouble())
        return NumberUtils.round(amount * factor, 0).toLong()
    }

    fun fromMinorUnits(minor: Long, code: String): Double? {
        val meta = lookup(code) ?: return null
        val factor = Math.pow(10.0, meta.exponent.toDouble())
        return minor / factor
    }

    private fun suffixLocale(locale: String): Boolean {
        val normalized = locale.lowercase()
        return normalized.startsWith("de") || normalized.startsWith("fr") ||
            normalized.startsWith("it") || normalized.startsWith("es")
    }

    fun formatCurrency(amount: Double, code: String, locale: String): String? {
        val meta = lookup(code) ?: return null
        val formatted = I18nUtils.formatNumberLocale(amount, locale, meta.exponent)
        return if (suffixLocale(locale)) "$formatted ${meta.symbol}" else "${meta.symbol}$formatted"
    }
}
